### PR TITLE
job file: improve style

### LINKED ISSUE #2150: Dump environment to stderr on job fail?
If an error is trapped by the job script, or when running in debug mode, we could this before exiting.
```
env 1>&2
```
@cylc/core - any objections to this?  The environment might occasionally have useful diagnostic info...

### PR DESCRIPTION
**Pull request by matthewrmshin**
Separate the job script into static part and generated part.
* Static part now lives in `lib/cylc/job.sh`, which will be loaded by the job file in run time.
* Generated part will continue to be generated by `cylc.job_file`. It is now quite small and contains:
  * The header, unchanged.
  * Essential environment variables to provide an initial environment, at the top.
  * Logic fragments in shell functions for cylc environment, user environment, user script, etc.
  * A line to load the static job script library, and a line to call its main function.
  * The footer, unchanged.

Reduce number of environment variables in job file.
* Mark old variables as duplicates.
* Move communication related variables to contact file.
* Remove variables that should not be needed by anything.

Improve diagnostics.
* Job script and `cylc message` no longer repeat one another.
* Report only relevant items for jobs. E.g. don't report suite host, port, etc.
* New `err-script` allows user defined scripts to run at the end of the error trap. Close #2150.

Quote and brace shell and environment variable values.
* Safer - in case there are spaces and other special characters in file system paths.
* Logic easier to read.

Remove space from keys in "job_conf".

(Rationale behind this change is that I want to review the logic for job script generation before I move on to look at #2043.)

### CODE REVIEW

**Comment by matthewrmshin** [comment]
(I'd also quite like to be able to separate the job script into little functions. Static parts can then reside somewhere in `CYLC_DIR` and sourced in by the generated job script at run time. The generated part of the job script should then become quite small, with only the essential parts visible to users.)

**Comment by hjoliver** [comment]
Brilliant :boom:

**Comment by matthewrmshin** [comment]
Please feel free to raise any concerns, objections, etc that you may feel with this change. In particular:
* Shell is now restricted to bash only.
* STDOUT for a job will look a bit different.
* Task will now fail if an existing suite has pre-script, script and post-script fragments that only work if they are concatenated together.
* Is there anything else that we ought to change or discard?
* Is it worth adding some syntax checking of the generated script (with `bash -n`)?

**Comment by matthewrmshin** [comment]
Having another thought about this.

> E.g.: if we have the start of a block in pre-script which only ends in post-script then the new logic will fail.

I have not heard of a use case for this, but real users often do the unexpected.

I have separated `pre-script`, `script` and `post-script` in 3 functions. There is no reason for doing this, except for making them stand alone as logic units. I can combine the 3 functions into one, so exotic logic like the above can continue to work.

> Is it worth adding bash syntax checking of the generated script (with `bash -n`)?

This is now essential because I have positioned error trapping in the static main function. This means bash syntax error in user scripts will not be trapped. A quick `bash -n` command on the job file after it is generated should not cost too much, but can be quite nice - so the job does not have to enter a batch system queue and then fail due to a bash syntax error. Alternatively,  if we are happy to keep `pre-script`, `script` and `post-script` separated, then we can in theory run `bash -n` on these settings at validation - which is more efficient for the suite - as we only have to do it once at start up.

**Comment by hjoliver** [comment]
> Please comment on preference.

(I still need to think about this, hopefully tomorrow...)

File: lib/cylc/job_file.py [hunk 1]
```diff
@@ -89,21 +84,32 @@ def write(self, local_job_file_path, job_conf):
         os.chmod(local_job_file_path, mode)
 
     @classmethod
```
Comment (matthewrmshin) [inline lib/cylc/job_file.py]: Done.

File: lib/cylc/job.sh [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+# Arguments:
+#   NAME - name of function without the "cylc::job::inst::" prefix
+# Returns:
+#   return 0, or the return code of the function if called
+cylc::job::run_inst_func() {
+    local NAME="$1"
+    if declare -F "cylc::job::inst::${NAME}" 1>'/dev/null' 2>&1; then
+        "cylc::job::inst::${NAME}"
+    fi
+}
```
Comment (hjoliver) [inline lib/cylc/job.sh]: Insert blank line, for consistency?

Comment (matthewrmshin) [inline lib/cylc/job.sh]: fixed.

**Comment by matthewrmshin** [comment]
(I have now added `bash -n` syntax checking at the point when the job file is generated.)

**Comment by hjoliver** [comment]
> Shell is now restricted to bash only.

I can still set `[[[job]]]shell=/bin/ksh`, so I guess that item (and its documentation) needs to be removed.  (BTW I've seen a lot of ksh scripts from the MO over the years - do your users not set that in their suites?)

**Comment by hjoliver** [comment]
> > E.g.: if we have the start of a block in pre-script which only ends in post-script then the new logic will fail.

> I have not heard of a use case for this, but real users often do the unexpected.

I can't think of a use case either, so I'm fine with this so long as we document that the script items have to be valid on their own.

**Comment by hjoliver** [comment]
> Alternatively, if we are happy to keep pre-script, script and post-script separated, then we can in theory run bash -n on these settings at validation - which is more efficient for the suite - as we only have to do it once at start up.

I like this idea, but perhaps only as a non-default option - it might really slow down validation of very large suites?

**Comment by hjoliver** [comment]
> This means bash syntax error in user scripts will not be trapped.

I was concerned about this, but it seems you do mean literally just *syntax* errors, which are now handled by `bash -n`; any command with non-zero exit status is still trapped as normal.

**Comment by matthewrmshin** [comment]
On ksh, we have advertised to our users that cylc job script should all be written in bash, so I hope we don't have any suites doing `[[[job]]]shell=/bin/ksh`. I'll remove reference to ksh from the docs.

On trapping syntax error... Yes, only bash syntax error is essential until the error trap is switched on - because everything is wrapped in functions, nothing actually gets executed until the main function is called. However, if bash fails to parse the script, it will not get to the main function.

**Comment by trwhitcomb** [comment]
ACK!  We still are using Korn shell, although most of the scripts should validate as such.

**Comment by hjoliver** [comment]
@matthewrmshin - looks like @trwhitcomb has thrown a spanner in the works!  Our own fault for support ksh in the first place I suppose.

Tim, if you look at a job script generated by a suite running with this branch, it's a beautiful thing - all the boiler-plate code is gone, leaving just what users want to see.  But as the job file becomes more sophisticated it gets harder to support both shells because they have less in common in terms of function syntax etc.  

If your users don't use screeds of inlined shell scripting (?do they?) it's very likely their tasks will still work as-is under bash.  What do you think - is this a justifiable change to force on users?

**Comment by matthewrmshin** [comment]
@trwhitcomb Thanks for the info. Some quick questions:
* (As @hjoliver already asked...) Is the usage of `ksh` in the job script just out of habit (like some of our users)? Or are they using some advanced ksh93-only syntax? If the former, the script is likely `sh`, `ksh88` and `bash` compatible and should just port to `bash` with no issue. The latter is trickier, but it should still be possible to make some simple change to port the logic into `bash`. Alternatively, one can always move the logic out into its own command in the suite's `bin/` directory. What do you think?
* What does `set -S` do? It does not appear to be documented in `man ksh`.

**Comment by matthewrmshin** [comment]
Branch rebased.

**Comment by trwhitcomb** [comment]
@matthewrmshin @hjoliver Just to clarify, I really like these changes for the reasons outlined above - makes the job script very much nicer and cleaner.

I'll need a little time to digest the implications of this, but I wanted to at least toss in the spanner to respond to the "nobody uses this anyway!" that I saw happening.  I suspect that it will end up being OK (as the Korn shell scripts would just end up being called from the resulting bash script) but I need to check a few things.

**Comment by matthewrmshin** [comment]
@trwhitcomb thanks for taking the time to look at this. Much appreciated.

**Comment by hjoliver** [comment]
@trwhitcomb - 

> ... OK (as the Korn shell scripts would just end up being called from the resulting bash script) ...

That would certainly be OK.  The only problem would be if users were in the habit of inlining significant amounts of ksh-specific scripting in the suite.rc `script` items, instead of calling an external script. 

@matthewrmshin - if the worst comes to the worst, would it be very hard to make this work under ksh too?  And perhaps then mark ksh support as deprecated as a warning prior to dropping it in the future?

**Comment by matthewrmshin** [comment]
> if the worst comes to the worst, would it be hard to make this work under ksh too?

Anything is possible. If we only need to consider ksh93 and not ksh88, I'll only need to change:
* `local` to `typeset` when declaring variables in functions.
* Function names can no longer use `::` as package delimiters. (This is really for readability only, so no big deal.) I'll use double underscore `__` instead.
* I'll leave PS4 prompt as default `+ ` in ksh.
* `typeset -f` instead of `declare -F` to check for existence of a function.

**Comment by matthewrmshin** [comment]
I have now brought back support for ksh (but only ksh93).

I have also introduced `err-script` to allow users to configure custom logic to call at the end of the error trap.

File: tests/jobscript/11-shell-ksh/suite.rc [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    [[reference test]]
+        expected task failures = foo.1
+    [[events]]
+        abort on stalled = True
+[scheduling]
+    [[dependencies]]
+        graph = foo
+[runtime]
+    [[foo]]
+        script = test "${CYLC_TASK_SUBMIT_NUMBER}" -ge 2
```
Comment (oliver-sanders) [inline tests/jobscript/11-shell-ksh/suite.rc]: Could potentially use some ksh only syntax e.g. `print`.

File: doc/suiterc.tex [hunk 4]
```diff
@@ -1132,25 +1132,42 @@ \subsection{[runtime]}
...
 \item {\em default:} (none)
 \item {\em example:} \lstinline@env-script = "echo Hello World"@
 \end{myitemize}
 
+\paragraph[err-script]{[runtime] \textrightarrow [[\_\_NAME\_\_]] \textrightarrow err-script}
+
+This is any custom script to be invoked at the end the error trap.
+The output of this will always be sent to STDERR and \lstinline=$1= is set to
+the name of the signal caught by the error trap. The script should fast and
+uses very little system resource to ensure that the error trap can return
```
Comment (matthewrmshin) [inline doc/suiterc.tex]: fixed.

File: tests/jobscript/11-shell-ksh/suite.rc (hunk 3 above)
Comment (matthewrmshin) [inline tests/jobscript/11-shell-ksh/suite.rc]: Done.

File: doc/cug.tex [hunk 5]
```diff
@@ -5114,30 +5141,57 @@ \section{Task Implementation}
...
+modification so long as they return standard exit status, i.e. zero on
+success and non-zero for failure, and do not internally spawn detaching
+processes (see~\ref{DetachingJobs}).
+
+\subsection{Task Job Scripts}
+\label{JobScripts}
+
+When a task is ready cylc generates a {\em task job script} that encapsulates
+the runtime settings (environment and scripts) defined for it in the suite.rc
+file. The job script is submitted to run by the {\em job submission method}
```
Comment (hjoliver) [inline doc/cug.tex]: This old terminology should be updated ("job submission method"  --> "batch system")

File: doc/cug.tex [hunk 6]
```diff
@@ -5114,30 +5141,57 @@ \section{Task Implementation}
...
-\lstinline=cylc task message "failed"= back to the suite. The shell
-\lstinline=set -e= option can be used in lieu of explicit error
-checks for every command:
+A task job script in cylc automatically traps ERR (for errors) and EXIT (for
+unexpected exits). When the trap is triggered, the task job script will send a
+failed message back to the suite before aborting.
+
+Therefore, commands in user scripts should abort with non-zero exit status if a
+fatal error occurs (this is just good coding practice anyway). To avoid a
+command that expects a non-zero return code from triggering a trap, you can use
```
Comment (hjoliver) [inline doc/cug.tex]: Probably should to "to *prevent* a command..."

File: doc/cug.tex [hunk 7]
```diff
@@ -5229,15 +5280,11 @@ \section{Task Job Submission and Management}
...
-\label{JobScripts}
-
-When a task is ready cylc generates a {\em task job script} that encapsulates
-the runtime settings (environment and scripting) defined for it in the suite.rc
-file. The job script is submitted to run by the {\em job submission method}
-chosen for the task. Different tasks can have different job submission methods.
-Like other runtime properties, you can set a suite default job submission
-method and override it for specific tasks or families:
+When a task is ready cylc generates a job script (see~\ref{JobScripts}). The
+job script is submitted to run by the {\em job submission method} chosen for
```
Comment (hjoliver) [inline doc/cug.tex]: "batch system"

File: doc/suiterc.tex [hunk 8]
```diff
@@ -1132,25 +1132,42 @@ \subsection{[runtime]}
...
 
 \begin{myitemize}
 \item {\em type:} string
 \item {\em default:} (none)
 \item {\em example:} \lstinline@env-script = "echo Hello World"@
 \end{myitemize}
 
+\paragraph[err-script]{[runtime] \textrightarrow [[\_\_NAME\_\_]] \textrightarrow err-script}
+
+This is any custom script to be invoked at the end the error trap.
```
Comment (hjoliver) [inline doc/suiterc.tex]: "... at the end *of* the error trap, if the task fails."?

File: doc/suiterc.tex [hunk 9]
```diff
@@ -1331,14 +1350,14 @@ \subsection{[runtime]}
...
-\lstinline@export item=value@ - valid for both bash and ksh
-because \lstinline=value= is entirely user-defined - but cylc would have
-to be modified slightly to allow use of the C shell.
+Location of the command used to interpret the job script submitted by the
+suite daemon when a task is ready to run. This can be set to the location of
+\lstinline=bash= in the job host if the shell is not installed in the standard
+location.
+{\em Note: It has no bearing on any sub-shells that may be called by the job script.}
+
+Setting this to the path of a ksh93 interpretter is supported but not recommended.
```
Comment (hjoliver) [inline doc/suiterc.tex]: One 't' in "interpreter".

Comment (hjoliver) [inline doc/suiterc.tex]: Shall we give a deprecation warning for ksh93 (here, and print one if detected on file parsing) - so we can drop support sometime in the future?

File: lib/cylc/job.sh [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+__OUT__
+    # Derived environment variables
+    export CYLC_SUITE_LOG_DIR="${CYLC_SUITE_RUN_DIR}/log/suite"
+    CYLC_SUITE_WORK_DIR_ROOT="${CYLC_SUITE_WORK_DIR_ROOT:-${CYLC_SUITE_RUN_DIR}}"
+    export CYLC_SUITE_SHARE_DIR="${CYLC_SUITE_WORK_DIR_ROOT}/share"
+    export CYLC_SUITE_WORK_DIR="${CYLC_SUITE_WORK_DIR_ROOT}/work"
+    export CYLC_TASK_CYCLE_POINT="$(cut -d '/' -f 1 <<<"${CYLC_TASK_JOB}")"
+    export CYLC_TASK_NAME="$(cut -d '/' -f 2 <<<"${CYLC_TASK_JOB}")"
+    # The "10#" part ensures that the submit number is interpretted in base 10.
+    # Otherwise, a zero padded number will be interpretted as an octal.
```
Comment (hjoliver) [inline lib/cylc/job.sh]: One 't' in 'interpreted'.

File: lib/cylc/job.sh [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+    rmdir "${CYLC_TASK_WORK_DIR}" 2>'/dev/null' || true
+    # Send task succeeded message
+    wait "${CYLC_TASK_MESSAGE_STARTED_PID}" 2>'/dev/null' || true
+    cylc task message 'succeeded' || true
+    trap '' EXIT
+    exit 0
+}
+
+###############################################################################
+# Run a function in the task job instant file, if possible.
```
Comment (hjoliver) [inline lib/cylc/job.sh]: "instance"?

**Comment by hjoliver** [comment]
A few minors doc and comment typos, otherwise this is all good from my perspective.

**Comment by matthewrmshin** [comment]
I have reinstated `CYLC_SUITE_HOST` and `CYLC_SUITE_OWNER`, but only if the contact file is available. These variables are used by various suites in our repositories.

**Comment by hjoliver** [comment]
@matthewrmshin -
> ... but only if the contact file is available

would you mind explaining that (so I don't have to figure it out)?

Also, could you list in one place any other of the old env vars removed?

**Comment by hjoliver** [comment]
A minimal local test suite on this branch now has task messaging failures:
```
ERROR: task messaging failure.
Traceback (most recent call last):
  File "/home/vagrant/cylc/bin/cylc-message", line 65, in main
    task_message.send(args)
  File "/home/vagrant/cylc/lib/cylc/task_message.py", line 79, in send
    SuiteSrvFilesManager().load_contact_file(self.suite))
  File "/home/vagrant/cylc/lib/cylc/suite_srv_files_mgr.py", line 359, in load_contact_file
    self.FILE_BASE_CONTACT, reg, owner, host, content=True)
  File "/home/vagrant/cylc/lib/cylc/suite_srv_files_mgr.py", line 230, in get_auth_item
    path = os.path.join(os.environ[key], self.DIR_BASE_SRV)
  File "/usr/lib/python2.7/UserDict.py", line 23, in __getitem__
    raise KeyError(key)
KeyError: 'CYLC_SUITE_RUN_DIR_ON_SUITE_HOST'
```

**Comment by hjoliver** [comment]
Here's a diff of `job.out` for same task on master vs this branch, the task has `script = printenv | grep CYLC`.

Questions: 

- are you sure that some of the removed "Suite and Task Identity" info isn't useful sometimes?
- ditto env vars (especially the one mentioned just above!)
```diff
1,11c1,3
< JOB SCRIPT STARTING
< cylc Suite and Task Identity:
<   Suite Name  : foo
<   Suite Host  : localhost
<   Suite Port  : 43076
<   Suite Owner : vagrant
<   Task ID     : foo.1
<   Task Host   : vagrant-ubuntu-trusty-64
<   Task Owner  : vagrant
<   Task Submit No.: 1
<   Task Try No.: 1
---
> Suite    : foo
> Task Job : 1/foo/01 (try 1)
> User@Host: vagrant@vagrant-ubuntu-trusty-64
14,17c6
< cylc (scheduler - 2017-02-17T22:18:45Z): started at 2017-02-17T22:18:45Z
< CYLC_SUITE_RUN_DIR_ON_SUITE_HOST=/home/vagrant/cylc-run/foo
< CYLC_TASK_SSH_LOGIN_SHELL=True
< CYLC_DEBUG=False
---
> 2017-02-17T22:17:44Z NORMAL - started
21d9
< CYLC_TASK_MSG_TIMEOUT=30.000000
29d16
< CYLC_MODE=scheduler
40d26
< CYLC_TASK_MSG_RETRY_INTVL=5.000000
43d28
< CYLC_JOB_PID=9475
45,46d29
< CYLC_DIR_ON_SUITE_HOST=/home/vagrant/cylc
< CYLC_TASK_MSG_MAX_TRIES=7
49,50d31
< CYLC_SUITE_PORT=43076
< CYLC_TASK_COMMS_METHOD=default
56,57c37,38
< cylc (scheduler - 2017-02-17T22:18:49Z): succeeded at 2017-02-17T22:18:49Z
< JOB SCRIPT EXITING (TASK SUCCEEDED)
---
> CYLC_TASK_JOB=1/foo/01
> 2017-02-17T22:17:54Z NORMAL - succeeded
```

**Comment by hjoliver** [comment]
To answer my own question, I think it is clear that none of those variables would be useful in user scripting (although perhaps occasionally useful diagnostic info?).  So just internal use of `CYLC_SUITE_RUN_DIR_ON_SUITE_HOST` is the problem.

**Comment by matthewrmshin** [comment]
Sorry. I'll try to fix the issue on Monday. I don't think things like port number needs to be in the job environment. They are now in the contact file. Other variables such as job pid are now in the job status file.

**Comment by matthewrmshin** [comment]
The contact file is now not written to the remote job host if communication method is "poll". It does not make sense to have a contact file when backward communication is not possible.

**Comment by matthewrmshin** [comment]
(Travis CI is happy again.)

**Comment by trwhitcomb** [comment]
Hi guys - sorry for the delay coming back around on this.  Was all set then we lost power for a few days!

I'm OK with this now that I've had a chance to look at what we're doing.  For posterity, here are the answers to questions from @matthewrmshin and why I think we're fine now:

1. Reason for using Korn shell for scripts is worse than habit - it's the policy decided on a long time ago.  Given the choices, I'm thankful they didn't go with c shell.  So, all our task scripts are written in Korn shell.  I suspect that about 99% of them will run just fine as Bash.

2. The "-S" I refer to is actually the PBS directive, telling it to interpret the task script as Korn shell instead.

I don't think we'll be able to convert all our scripts over to Bash due to the aforementioned policy.  However, almost all our tasks have a "script" command that just points out to run another script, so you'd just be in the situation where the Bash script written out by cylc will call a Korn shell script written by us.  Any inline scripting we have looks to be bash and Korn compatible.

Nice work, guys!

**Comment by matthewrmshin** [comment]
@trwhitcomb you should now be okay as we have brought back ksh support (ksh93 only, however). I can understand policy like what you have described. It hasn't been long since we moved away, and there are still many users uncomfortable with the idea of moving away from ksh.

**Pull request merged**